execution/block-partitioner/src/v2/mod.rs

```diff
@@ -0,0 +1,719 @@
+                            });
+                    });
+            });
+        });
+        let duration = timer.stop_and_record();
+        info!("new_tid_table={duration}");
+        (txn_idx_matrix, start_index_matrix, finalized_indexs)
+    }
+}
+
+impl BlockPartitioner for PartitionerV2 {
+    fn partition(
+        &self,
+        txns: Vec<AnalyzedTransaction>,
+        num_executor_shards: usize,
+    ) -> Vec<SubBlocksForShard<AnalyzedTransaction>> {
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["preprocess"])
+            .start_timer();
+        let num_txns = txns.len();
+        let num_senders = AtomicUsize::new(0);
+        let num_keys = AtomicUsize::new(0);
+        let mut senders: Vec<RwLock<Option<SenderIdx>>> = Vec::with_capacity(num_txns);
+        let mut wsets: Vec<RwLock<HashSet<StorageKeyIdx>>> = Vec::with_capacity(num_txns);
+        let mut rsets: Vec<RwLock<HashSet<StorageKeyIdx>>> = Vec::with_capacity(num_txns);
+        let sender_idx_table: DashMap<Sender, SenderIdx> =
+            DashMap::with_shard_amount(self.dashmap_num_shards);
+        let key_idx_table: DashMap<StateKey, StorageKeyIdx> =
+            DashMap::with_shard_amount(self.dashmap_num_shards);
+        let trackers: DashMap<StorageKeyIdx, RwLock<ConflictingTxnTracker>> =
+            DashMap::with_shard_amount(self.dashmap_num_shards);
+        for txn in txns.iter() {
+            senders.push(RwLock::new(None));
+            wsets.push(RwLock::new(HashSet::with_capacity(txn.write_hints().len())));
+            rsets.push(RwLock::new(HashSet::with_capacity(txn.read_hints().len())));
+        }
+        let timer_1 = MISC_TIMERS_SECONDS
+            .with_label_values(&["preprocess__main"])
+            .start_timer();
+        self.thread_pool.install(|| {
+            (0..num_txns)
+                .into_par_iter()
+                .for_each(|txn_idx: OriginalTxnIdx| {
+                    let txn = &txns[txn_idx];
+                    let sender = txn.sender();
+                    let sender_idx = *sender_idx_table
+                        .entry(sender)
+                        .or_insert_with(|| num_senders.fetch_add(1, Ordering::SeqCst));
+                    *senders[txn_idx].write().unwrap() = Some(sender_idx);
+                    let num_writes = txn.write_hints().len();
+                    for (i, storage_location) in txn
+                        .write_hints()
+                        .iter()
+                        .chain(txn.read_hints().iter())
+                        .enumerate()
+                    {
+                        let key = storage_location.state_key().clone();
+                        let key_idx = *key_idx_table
+                            .entry(key)
+                            .or_insert_with(|| num_keys.fetch_add(1, Ordering::SeqCst));
+                        let is_write = i < num_writes;
+                        if is_write {
```

Comment:
This is a bit confusing.. why not iterate through read_set and write_set separately?